presto-main/src/main/java/com/facebook/presto/event/QueryMonitor.java

```diff
@@ -273,6 +281,31 @@ private QueryContext createQueryContext(SessionRepresentation session, Optional<
         return Optional.empty();
     }
 
+    private Optional<String> createJsonQueryPlan(QueryInfo queryInfo)
+    {
+        try {
+            if (queryInfo.getOutputStage().isPresent()) {
+                JsonCodec<Map<PlanFragmentId, JsonPlanFragment>> codec = JsonCodec.mapJsonCodec(PlanFragmentId.class, JsonPlanFragment.class);
+                ImmutableMap.Builder<PlanFragmentId, JsonPlanFragment> fragmentJsonMap = ImmutableMap.builder();
+                for (StageInfo stage : getAllStages(queryInfo.getOutputStage())) {
+                    PlanFragmentId fragmentId = stage.getPlan().getId();
+                    if (!stage.getPlan().getJsonRepresentation().isPresent()) {
+                        log.warn("Error creating json plan for query %s: json representation is not present for fragment %s", queryInfo.getQueryId(), fragmentId);
+                        return Optional.empty();
+                    }
+                    JsonPlanFragment jsonPlanFragment = new JsonPlanFragment(stage.getPlan().getJsonRepresentation().get());
+                    fragmentJsonMap.put(fragmentId, jsonPlanFragment);
+                }
+                return Optional.of(codec.toJson(fragmentJsonMap.build()));
+            }
+        }
+        catch (Exception e) {
+            // Don't fail to create event if the plan can not be created, e.g. due to a json error
```

Comment:
I am still confused by this - why would there be a JSON error here? I think it's OK to be safe, but I wouldn't explicitly call out JSON because it's misleading.